### PR TITLE
v1.5 logging (version 2)

### PR DESCRIPTION
**Pull request by Aaronontheweb**
## Changes

Fixes https://github.com/akkadotnet/akka.net/pull/6394

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Design discussion issue https://github.com/akkadotnet/akka.net/pull/6394#issuecomment-1424987909
* [x] Changes in public API reviewed, if any.
* [x] I have added website documentation for this feature.

### Latest `dev` Benchmarks 

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2486/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-WCODYP : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|                      Method |     Mean |    Error |   StdDev |   Median | Ratio | RatioSD |   Gen0 |   Gen1 |   Gen2 | Allocated | Alloc Ratio |
|---------------------------- |---------:|---------:|---------:|---------:|------:|--------:|-------:|-------:|-------:|----------:|------------:|
|             LogInfoNoParams | 163.2 ns |  3.25 ns |  8.22 ns | 160.3 ns |  1.00 |    0.00 | 0.0110 | 0.0060 | 0.0010 |      64 B |        1.00 |
|              LogInfo1Params | 383.5 ns | 19.88 ns | 58.61 ns | 401.0 ns |  2.36 |    0.39 | 0.0250 | 0.0120 |      - |     160 B |        2.50 |
|              LogInfo2Params | 401.8 ns |  9.62 ns | 28.21 ns | 410.3 ns |  2.47 |    0.22 | 0.0300 | 0.0150 |      - |     192 B |        3.00 |
|        LogInfoWithException | 164.6 ns |  3.26 ns |  7.62 ns | 160.9 ns |  1.01 |    0.07 | 0.0100 | 0.0050 |      - |      64 B |        1.00 |
| LogInfoWithException1Params | 383.3 ns | 20.53 ns | 60.52 ns | 398.9 ns |  2.36 |    0.38 | 0.0260 | 0.0130 | 0.0010 |     160 B |        2.50 |
| LogInfoWithException2Params | 408.3 ns |  9.78 ns | 28.82 ns | 416.9 ns |  2.51 |    0.22 | 0.0300 | 0.0150 |      - |     192 B |        3.00 |

### This PR's Benchmarks

Include data from after this change here.

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
Going to run some benchmark numbers before I start calling the generic methods we support

**Comment by Aaronontheweb** [comment]
Current version's benchmarks - which is more or less performing the same work under the hood that the `dev` branch is doing:

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2604/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-KJHKWD : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|                      Method |     Mean |    Error |   StdDev |   Median | Ratio | RatioSD |   Gen0 |   Gen1 |   Gen2 | Allocated | Alloc Ratio |
|---------------------------- |---------:|---------:|---------:|---------:|------:|--------:|-------:|-------:|-------:|----------:|------------:|
|             LogInfoNoParams | 169.9 ns |  3.39 ns |  7.73 ns | 166.5 ns |  1.00 |    0.00 | 0.0100 | 0.0050 |      - |      64 B |        1.00 |
|              LogInfo1Params | 389.3 ns | 20.87 ns | 61.53 ns | 388.4 ns |  2.30 |    0.38 | 0.0250 | 0.0120 |      - |     160 B |        2.50 |
|              LogInfo2Params | 408.1 ns |  8.14 ns | 23.86 ns | 412.6 ns |  2.41 |    0.19 | 0.0310 | 0.0160 | 0.0010 |     192 B |        3.00 |
|        LogInfoWithException | 169.1 ns |  3.38 ns |  8.47 ns | 164.6 ns |  1.00 |    0.08 | 0.0110 | 0.0060 | 0.0010 |      64 B |        1.00 |
| LogInfoWithException1Params | 392.6 ns | 20.44 ns | 60.28 ns | 397.0 ns |  2.32 |    0.37 | 0.0260 | 0.0130 | 0.0010 |     160 B |        2.50 |
| LogInfoWithException2Params | 409.2 ns |  8.17 ns | 23.45 ns | 423.7 ns |  2.41 |    0.18 | 0.0310 | 0.0160 | 0.0010 |     192 B |        3.00 |

**Comment by Aaronontheweb** [comment]
Changed all method calls to use our new generic overloads (I'll need to compare before / after .DLL sizes to make sure we haven't done anything evil there) - performance for logging calls is basically now bounded at 260ish nanoseconds for up to 6 parameters.

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2604/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-PCKFPT : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|                      Method |     Mean |   Error |   StdDev |   Median | Ratio | RatioSD |   Gen0 |   Gen1 |   Gen2 | Allocated | Alloc Ratio |
|---------------------------- |---------:|--------:|---------:|---------:|------:|--------:|-------:|-------:|-------:|----------:|------------:|
|             LogInfoNoParams | 169.9 ns | 3.36 ns |  7.31 ns | 166.5 ns |  1.00 |    0.00 | 0.0110 | 0.0060 | 0.0010 |      64 B |        1.00 |
|              LogInfo1Params | 260.7 ns | 7.95 ns | 23.44 ns | 262.0 ns |  1.54 |    0.16 | 0.0170 | 0.0090 | 0.0010 |     104 B |        1.62 |
|              LogInfo2Params | 261.1 ns | 7.81 ns | 23.02 ns | 258.5 ns |  1.54 |    0.15 | 0.0170 | 0.0090 | 0.0010 |     104 B |        1.62 |
|        LogInfoWithException | 164.6 ns | 3.29 ns |  7.22 ns | 160.7 ns |  0.97 |    0.03 | 0.0100 | 0.0050 |      - |      64 B |        1.00 |
| LogInfoWithException1Params | 266.7 ns | 8.59 ns | 25.34 ns | 277.1 ns |  1.58 |    0.17 | 0.0170 | 0.0090 | 0.0010 |     104 B |        1.62 |
| LogInfoWithException2Params | 268.4 ns | 8.34 ns | 24.58 ns | 277.8 ns |  1.60 |    0.16 | 0.0170 | 0.0090 | 0.0010 |     104 B |        1.62 |

**Comment by Aaronontheweb** [comment]
Moved all generic methods out of the `ILoggingAdapter` interface and onto extension methods instead. Simplified the surface area of the `ILoggingAdapter` significantly so it can be extended in the next phase of this logging overhaul. Had to expose the `ILogFormatter` on the API in order to do this.

Now down to about 240 ns.

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2604/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-MAIQQP : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|                      Method |     Mean |   Error |   StdDev |   Median | Ratio | RatioSD |   Gen0 |   Gen1 |   Gen2 | Allocated | Alloc Ratio |
|---------------------------- |---------:|--------:|---------:|---------:|------:|--------:|-------:|-------:|-------:|----------:|------------:|
|             LogInfoNoParams | 163.6 ns | 3.24 ns |  7.70 ns | 159.3 ns |  1.00 |    0.00 | 0.0100 | 0.0050 |      - |      64 B |        1.00 |
|              LogInfo1Params | 243.7 ns | 8.28 ns | 24.42 ns | 245.5 ns |  1.50 |    0.17 | 0.0170 | 0.0090 | 0.0010 |     104 B |        1.62 |
|              LogInfo2Params | 237.8 ns | 8.40 ns | 24.76 ns | 247.3 ns |  1.45 |    0.16 | 0.0170 | 0.0090 | 0.0010 |     104 B |        1.62 |
|        LogInfoWithException | 164.3 ns | 3.28 ns |  7.26 ns | 160.4 ns |  1.01 |    0.06 | 0.0100 | 0.0050 |      - |      64 B |        1.00 |
| LogInfoWithException1Params | 245.2 ns | 8.33 ns | 24.55 ns | 256.2 ns |  1.51 |    0.17 | 0.0170 | 0.0090 | 0.0010 |     104 B |        1.62 |
| LogInfoWithException2Params | 246.5 ns | 8.05 ns | 23.73 ns | 254.0 ns |  1.51 |    0.17 | 0.0170 | 0.0090 | 0.0010 |     104 B |        1.62 |

**Comment by Aaronontheweb** [comment]
@Arkatufus does this mean I killed the multi-node testkit?

> [ERROR][02/16/2023 04:11:46.329][Thread 0015][akka://TestRunnerLogging/user/TcpLogger] Method not found: 'Void Akka.Event.ILoggingAdapter.Info(System.String, System.Object[])'.

File: src/core/Akka/Event/ILoggingAdapter.cs [hunk 1]
```diff
@@ -6,246 +6,748 @@
...
-        public void Debug(Exception cause, string format, params object[] args){ }
+        void Log(LogLevel logLevel, Exception cause, string format);
 
-        /// <summary>
-        /// Logs a <see cref="LogLevel.InfoLevel" /> message.
-        /// </summary>
-        /// <param name="format">The message that is being logged.</param>
-        /// <param name="args">An optional list of items used to format the message.</param>
-        public void Info(string format, params object[] args) { }
+        void Log(LogLevel logLevel, Exception cause, LogMessage message);
```
Comment (Aaronontheweb) [inline src/core/Akka/Event/ILoggingAdapter.cs:682]: `ILoggingAdapter` only supports two direct logging calls now:

```csharp
void Log(LogLevel logLevel, Exception cause, LogMessage message);
void Log(LogLevel logLevel, Exception cause, string format);
```

Everything else is inputted via the extension methods earlier in the file.

File: src/core/Akka/Util/Reflection/ExpressionExtensions.cs [hunk 2]
```diff
@@ -110,12 +110,11 @@ internal static class ExpressionBasedParser
                         }
                     }
                 }
-                catch (Exception exception)
```
Comment (Aaronontheweb) [inline src/core/Akka/Util/Reflection/ExpressionExtensions.cs:113]: Fixed some compilation warnings here.

**Review by Aaronontheweb — COMMENTED** [review]
Detailed my changes, their rationales, and their benefits.

**Comment by Aaronontheweb** [comment]
Minus the issues with the MNTR, test suite passes.

I'm going to need to fix a few other issues found in this PR as well.

As for documentation - we'll need to document these changes in the https://github.com/akkadotnet/akka.net/blob/dev/docs/community/whats-new/akkadotnet-v1.5-upgrade-advisories.md area

File: src/core/Akka.Streams/Stage/Stage.cs [hunk 3]
```diff
@@ -72,7 +72,7 @@ public interface IStage<in TIn, out TOut> { }
     /// <see cref="IContext.AbsorbTermination"/> which stops the propagation of the termination signal, and puts the stage in a
     /// <see cref="IContext.IsFinishing"/> state. Depending on whether the stage has a pending pull signal it
     /// has not yet "consumed" by a push its <see cref="AbstractStage{TIn,TOut}.OnPull"/> handler might be called immediately or later. From
-    /// <see cref="AbstractStage{TIn,TOut}.OnPull"/> final elements can be pushed before completing downstream with <see cref="IContext.Finish"/> or
+    /// <see cref="AbstractStage{TIn,TOut}.OnPull"/> final elements can be pushed before completing downstream with <see cref="StatefulStage.Finish"/> or
```
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/core/Akka.Streams/Stage/Stage.cs:75]: actually, no - this change is accurate.

File: src/core/Akka.Streams/Stage/Stage.cs [hunk 4]
```diff
@@ -56,7 +56,7 @@ public interface IStage<in TIn, out TOut> { }
     /// <see cref="PushStage{TIn,TOut}"/> instead of <see cref="PushPullStage{TIn,TOut}"/>.
     /// </para>
     /// <para>
-    /// Stages are allowed to do early completion of downstream and cancel of upstream. This is done with <see cref="IContext.Finish"/>,
+    /// Stages are allowed to do early completion of downstream and cancel of upstream. This is done with <see cref="StatefulStage.Finish"/>,
```
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/core/Akka.Streams/Stage/Stage.cs:59]: actually, no - this change is accurate.

File: src/core/Akka.Streams/Stage/AbstractStage.cs [hunk 5]
```diff
@@ -440,7 +441,7 @@ public virtual void PreStart(ILifecycleContext context)
         /// with <see cref="IContext.IsFinishing"/>.
         /// </para>
         /// <para>
-        /// By default the finish signal is immediately propagated with <see cref="IContext.Finish"/>.
+        /// By default the finish signal is immediately propagated with <see cref="StatefulStage.Finish"/>.
```
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/core/Akka.Streams/Stage/AbstractStage.cs:444]: actually, no - this change is accurate.

File: src/core/Akka/Event/ILoggingAdapter.cs [hunk 6]
```diff
@@ -6,246 +6,748 @@
...
-        /// <param name="format">The message that is being logged.</param>
-        /// <param name="args">An optional list of items used to format the message.</param>
-        public void Debug(string format, params object[] args) { }
+        // /// <summary>
+        // /// Logs a message with a specified level.
+        // /// </summary>
+        // /// <param name="logLevel">The level used to log the message.</param>
+        // /// <param name="format">The message that is being logged.</param>
+        // /// <param name="args">An optional list of items used to format the message.</param>
+        // void Log(LogLevel logLevel, string format, params object[] args);
```
Comment (Aaronontheweb) [inline src/core/Akka/Event/ILoggingAdapter.cs:671]: Need to delete this (can do in later PR)

**Review by Arkatufus — APPROVED** [review]
LGTM

Microsoft does have their own class called `FormattableString` for these, but I think its too convoluted because it has to handle culture as well. This implementation should be better for our use case.

**Comment by Aaronontheweb** [comment]
> Microsoft does have their own class called `FormattableString` for these, but I think its too convoluted because it has to handle culture as well. This implementation should be better for our use case.

When we start work on 1.6, at which point I'm thinking we'll ditch .NET Standard 2.0 and go full-bore on .NET 7/8, we can start taking advantage of some of these newer APIs.

**Comment by Aaronontheweb** [comment]
Ok, going nuclear on API changes for Akka.NET v1.5 starting now ;)

**Pull request merged**